File: app/src/main/java/uk/co/victoriajanedavis/chatapp/data/repositories/cache/FriendshipCache.java
package uk.co.victoriajanedavis.chatapp.data.repositories.cache;

import java.util.List;
import java.util.UUID;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.annotations.NonNull;
import io.reactivex.annotations.Nullable;
import uk.co.victoriajanedavis.chatapp.injection.scopes.ApplicationScope;
import uk.co.victoriajanedavis.chatapp.data.model.db.FriendshipDbModel;
import uk.co.victoriajanedavis.chatapp.data.room.ChatAppDatabase;
import uk.co.victoriajanedavis.chatapp.data.room.daos.FriendshipDao;
import uk.co.victoriajanedavis.chatapp.domain.Cache.DiskCache;

@ApplicationScope
public class FriendshipCache implements DiskCache<UUID, FriendshipDbModel> {

    private final FriendshipDao dao;


    @Inject
    public FriendshipCache(ChatAppDatabase database) {
        this.dao = database.friendshipDao();
    }

    @Override
    public void putSingular(@NonNull FriendshipDbModel friendshipDbModel) {
        dao.insertFriendship(friendshipDbModel);
    }

    @Override
    public void putAll(@NonNull List<FriendshipDbModel> friendshipDbModels) {
        dao.insertFriendships(friendshipDbModels);
    }

    @Override
    public void replaceAll(@Nullable UUID uuid,
                           @NonNull List<FriendshipDbModel> friendshipDbModels) {
        dao.replaceAllAcceptedFriendships(friendshipDbModels);
    }

    @Override
    public void delete(@NonNull FriendshipDbModel friendshipDbModel) {
        dao.deleteFriendship(friendshipDbModel);
    }

    @Override
    public void clear() {
        dao.clear();
    }

    @NonNull
    @Override
    public Observable<FriendshipDbModel> getSingular(@NonNull UUID chat_uuid) {
        return dao.getFriendshipByChatUuid(chat_uuid).toObservable();
    }

    @NonNull
    @Override
    public Observable<List<FriendshipDbModel>> getAll(@Nullable UUID uuid) {
        return dao.getAllAcceptedFriendships().toObservable();
    }
}

File: app/src/main/java/uk/co/victoriajanedavis/chatapp/data/repositories/cache/FriendshipCache.kt
package uk.co.victoriajanedavis.chatapp.data.repositories.cache

import java.util.UUID

import javax.inject.Inject

import io.reactivex.Observable
import uk.co.victoriajanedavis.chatapp.injection.scopes.ApplicationScope
import uk.co.victoriajanedavis.chatapp.data.model.db.FriendshipDbModel
import uk.co.victoriajanedavis.chatapp.data.room.ChatAppDatabase
import uk.co.victoriajanedavis.chatapp.domain.Cache.DiskCache

@ApplicationScope
class FriendshipCache @Inject constructor(
    database: ChatAppDatabase
) : DiskCache<UUID, FriendshipDbModel> {

    private val dao = database.friendshipDao()


    override fun putSingular(value: FriendshipDbModel) {
        dao.insertFriendship(value)
    }

    override fun putAll(valuesList: List<FriendshipDbModel>) {
        dao.insertFriendships(valuesList)
    }

    override fun replaceAll(key: UUID?, valuesList: List<FriendshipDbModel>) {
        dao.replaceAllAcceptedFriendships(valuesList)
    }

    override fun delete(value: FriendshipDbModel) {
        dao.deleteFriendship(value)
    }

    override fun clear() {
        dao.clear()
    }

    // Key should be chatUuid
    override fun getSingular(key: UUID?): Observable<FriendshipDbModel> {
        return dao.getFriendshipByChatUuid(key!!).toObservable()
    }

    override fun getAll(key: UUID?): Observable<List<FriendshipDbModel>> {
        return dao.allAcceptedFriendships.toObservable()
    }
}
